### PR TITLE
Hia 292 add sentence length and terms to person sentences

### PR DESCRIPTION
**Pull request by popey2700**
#### Work done
- Added terms array to Sentence object
- Added mapping functions for both Nomis and nDelius to tranform to terms

#### Tests
Does this PR include tests for the changes introduced?
- [x] Yes
- [ ] No, I found it difficult to test
- [ ] No, they are not required for this change

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/Term.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.justice.digital.hmpps.hmppsintegrationapi.models
+
+data class Term(
+  val years: Int? = null,
+  val months: Int? = null,
+  val weeks: Int? = null,
+  val days: Int? = null,
```
[the remark this answers is not in the document]
Comment (luke-a-williams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/Term.kt:7]: thus they are either present and populated or not present... is defaulting to null the most sensible option here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/ndelius/Sentence.kt [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+  val length: Int? = null,
+  val lengthUnits: String? = null,
+) {
+  fun toTerm(): List<IntegrationApiTerm> {
+    return when (this.lengthUnits?.lowercase()) {
+      "years" -> listOf(IntegrationApiTerm(years = this.length))
+      "months" -> listOf(IntegrationApiTerm(months = this.length))
+      "weeks" -> listOf(IntegrationApiTerm(weeks = this.length))
+      "days" -> listOf(IntegrationApiTerm(days = this.length))
+      else -> listOf(IntegrationApiTerm())
```
Comment (luke-a-williams) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/ndelius/Sentence.kt:16]: The example on their swagger spec suggests that "Hours" is also possible. How are going to manage this?
Currently it looks like it will fall in the "else" category, and meaning that we have changed the time of sentence from something in hours to something that's not populated.
https://external-api-and-delius-dev.hmpps.service.justice.gov.uk/swagger-ui/index.html

Comment (luke-a-williams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/ndelius/Sentence.kt:16]: Further, what are all the possibilities of lengthUnits here? Can you link me to a source?

Comment (luke-a-williams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/ndelius/Sentence.kt:16]: Domain analysis yields the following from PreProd
`[
  {
    "domain": "COND_UNIT",
    "code": "DAYS",
    "description": "Days",
    "activeFlag": "Y",
    "listSeq": 2,
    "systemDataFlag": "Y",
    "subCodes": []
  },
  {
    "domain": "COND_UNIT",
    "code": "HOURS",
    "description": "Hours",
    "activeFlag": "Y",
    "listSeq": 1,
    "systemDataFlag": "Y",
    "subCodes": []
  },
  {
    "domain": "COND_UNIT",
    "code": "MONTHS",
    "description": "Months",
    "activeFlag": "Y",
    "listSeq": 4,
    "systemDataFlag": "Y",
    "subCodes": []
  },
  {
    "domain": "COND_UNIT",
    "code": "WEEKS",
    "description": "Weeks",
    "activeFlag": "Y",
    "listSeq": 3,
    "systemDataFlag": "Y",
    "subCodes": []
  },
  {
    "domain": "COND_UNIT",
    "code": "YEARS",
    "description": "Years",
    "activeFlag": "Y",
    "listSeq": 5,
    "systemDataFlag": "Y",
    "subCodes": []
  }
]`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/nomis/Term.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  val years: Int? = null,
+  val months: Int? = null,
+  val weeks: Int? = null,
+  val days: Int? = null,
+) {
+  fun toTerm(): IntegrationApiTerm = IntegrationApiTerm(
+    years = if (this.years != 0) this.years else null,
+    months = if (this.months != 0) this.months else null,
+    weeks = if (this.weeks != 0) this.weeks else null,
+    days = if (this.days != 0) this.days else null,
```
Comment (luke-a-williams) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/nomis/Term.kt:15]: Noticing that this function returns else null ... is the best decision we can make?
As pointed out earlier, the data coming from Nomis simply doesn't include these fields if they were null at Nomis end... can we implement this instead?
They achieve it with annotations so how would we achieve it?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt [hunk 4]
```diff
@@ -77,15 +110,27 @@ internal class SentencesControllerTest(
           [
             {
               "dateOfSentencing": "1990-01-01",
-              "isActive": true
+              "isActive": true,
+              "terms": [
+                { "years": 10, "months": null, "weeks": 5, "days": 5 },
+                { "years": 25, "months": null, "weeks": 5, "days": 4 }
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt:116]: Pointing out that this is NOT how the data is exposed to us from upstream - there are never null fields being sent to us, they are either missing or populated.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt [hunk 5]
```diff
@@ -77,15 +110,27 @@ internal class SentencesControllerTest(
...
+              ]
             },
             {
               "dateOfSentencing": "1991-01-01",
-              "isActive": false
+              "isActive": false,
+              "terms": [
+                { "years": 15, "months": 6, "weeks": 2, "days": null },
+                { "years": null, "months": 6, "weeks": 2, "days": 5 }
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt:124]: Pointing out that this is NOT how the data is exposed to us from upstream - there are never null fields being sent to us, they are either missing or populated.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt [hunk 6]
```diff
@@ -77,15 +110,27 @@ internal class SentencesControllerTest(
...
+              ]
             },
             {
               "dateOfSentencing": "1992-01-01",
-              "isActive": null
+              "isActive": null,
+              "terms": [
+                { "years": 15, "months": null, "weeks": null, "days": null },
+                { "years": null, "months": null, "weeks": 2, "days": null }
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/controllers/v1/SentencesControllerTest.kt:132]: Pointing out that this is NOT how the data is exposed to us from upstream - there are never null fields being sent to us, they are either missing or populated.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/nomis/PersonSentencesTest.kt [hunk 7]
```diff
@@ -22,6 +24,44 @@ class PersonSentencesTest : DescribeSpec(
...
+              days = 2,
+            ),
+            NomisTerm(
+              years = 7,
+              months = 3,
+              weeks = 4,
+              days = 0,
+            ),
+          ),
+        )
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/models/nomis/PersonSentencesTest.kt:43]: You need to have a test case where you test NomisTerms where any combination of those fields are missing.
I.e. years = 3 in one, months = 2 in the other.
You also need a case where you test a completely empty NomisTerm (because it is possible for no terms to exist)
You also need a case where you test a single Term in a list, to ensure smooth functionality for the singular case.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/services/GetSentencesForPersonServiceTest.kt [hunk 8]
```diff
@@ -93,13 +118,29 @@ internal class GetSentencesForPersonServiceTest(
...
               dateOfSentencing = LocalDate.parse("2004-01-01"),
               isActive = false,
+              terms = listOf(
+                IntegrationApiTerm(
+                  weeks = 18,
+                ),
+                IntegrationApiTerm(
+                  months = 11,
+                ),
+              ),
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/services/GetSentencesForPersonServiceTest.kt:143]: Why are you returning IntegrationApiSentence models whenever you are getting sentence information from nDelius? Surely you should be responding with an nDelius model constructor?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/smoke/person/SentencesSmokeTest.kt [hunk 9]
```diff
@@ -36,11 +36,27 @@ class SentencesSmokeTest : DescribeSpec(
...
             {
               "dateOfSentencing": "2019-08-24",
-              "isActive": true
+              "isActive": true,
+              "terms": [
+                {
+                  "years": null,
+                  "months": null,
+                  "weeks": null,
+                  "days": null
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/smoke/person/SentencesSmokeTest.kt:57]: Whilst this is true for right now, if there is no sentence term information why are we returning a bunch of nulls?
We are technically CREATING data when there isn't any by declaring it null. Instead this should match the upstream functionality and return an empty list when there are no terms.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/smoke/person/SentencesSmokeTest.kt [hunk 10]
```diff
@@ -36,11 +36,27 @@ class SentencesSmokeTest : DescribeSpec(
...
             {
               "dateOfSentencing": "2019-08-24",
-              "isActive": null
+              "isActive": null,
+              "terms": [
+                {
+                  "years": 1,
+                  "months": 2,
+                  "weeks": 3,
+                  "days": 4
```
Comment (luke-a-williams) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsintegrationapi/smoke/person/SentencesSmokeTest.kt:45]: I don't understand this smoke test, why would we have 'null' in the isActive part yet data in the terms?

**Review by luke-a-williams — CHANGES_REQUESTED** [review]
its taken me a while to go through this so it's likely I've probably missed things. I think we should call to discuss some issues in the data model.

**Review by luke-a-williams — APPROVED** [review]
Love the model and test improvements - Approving under the agreed definition to get a model in for testing. The HMPPS model might be subject to change following conversations with the full team early next week.
